File: game_scripts/levels/tests/teleporter_test.lua
--[[ Copyright (C) 2018 Google Inc.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
]]

local events = require 'dmlab.system.events'
local game = require 'dmlab.system.game'
local maze_generation = require 'dmlab.system.maze_generation'
local custom_entities = require 'common.custom_entities'
local make_map = require 'common.make_map'
local custom_observations = require 'decorators.custom_observations'
local timeout = require 'decorators.timeout'

local api = {}
local MAP_ENTITIES = '*********\n' ..
                     '* P * t *\n' ..
                     '*   *   *\n' ..
                     '* T *   *\n' ..
                     '*********\n'

function api:init(params)
  make_map.seedRng(1)
  local maze = maze_generation.mazeGeneration{entity = MAP_ENTITIES}
  api._map = make_map.makeMap{
      mapName = "teleport_test",
      mapEntityLayer = MAP_ENTITIES,
      useSkybox = true,
      callback = function (i, j, c, maker)
        if c == 'T' then
          return custom_entities.makeTeleporter(
              {maze:toWorldPos(i + 1, j + 1)},
              'teleporter_dest_second_room')
        end
        if c == 't' then
          return custom_entities.makeTeleporterTarget(
              {maze:toWorldPos(i + 1, j + 1)},
              'teleporter_dest_second_room')
        end
      end
  }
end

function api:hasEpisodeFinished(time)
  if game:playerInfo().teleported then
    events:add("PLAYER_TELEPORTED")
  end
  return false
end

function api:nextMap()
  return self._map
end

function api:updateSpawnVars(spawnVars)
  if spawnVars.classname == "info_player_start" then
    -- Spawn facing South.
    spawnVars.angle = "-90"
    spawnVars.randomAngleRange = "0"
  end
  return spawnVars
end

timeout.decorate(api, 60 * 60)
custom_observations.decorate(api)

return api

File: python/tests/teleporter_test.py
# Copyright 2018 Google Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import os
from absl.testing import absltest
import numpy as np
import six

import deepmind_lab


class TeleporterTest(absltest.TestCase):

  def test_movement(self):
    fps = 60
    env = deepmind_lab.Lab(
        'tests/teleporter_test', [
            'VEL.TRANS',
            'DEBUG.POS.TRANS',
            'DEBUG.POS.ROT',
        ],
        config={
            'fps': str(fps),
            'width': '80',
            'height': '80',
        })

    action_spec = env.action_spec()
    action_index = {action['name']: i for i, action in enumerate(action_spec)}

    action = np.zeros([len(action_spec)], dtype=np.intc)

    env.reset()
    vel = env.observations()['VEL.TRANS']
    self.assertTrue(np.array_equal(vel, np.array([0, 0, 0])))

    # Agent begins facing south
    initial_facing = env.observations()['DEBUG.POS.ROT']
    self.assertTrue(np.allclose(initial_facing, np.array([0, -90, 0]),
                                atol=0.1))

    # Player moves straight ahead through the teleporter
    action[action_index['MOVE_BACK_FORWARD']] = 1
    self.assertEqual(env.events(), [])
    for _ in six.moves.range(120):
      p_before = env.observations()['DEBUG.POS.TRANS']
      env.step(action, 1)
      p_after = env.observations()['DEBUG.POS.TRANS']
      if p_after[1] - p_before[1] > 100:
        break
    else:
      self.fail('Failed to teleport')
    self.assertEqual(env.events(), [('PLAYER_TELEPORTED', [])])
    env.step(action, 1)
    self.assertEqual(env.events(), [])

if __name__ == '__main__':
  if 'TEST_SRCDIR' in os.environ:
    deepmind_lab.set_runfiles_path(
        os.path.join(os.environ['TEST_SRCDIR'],
                     'org_deepmind_lab'))
  absltest.main()
